materialize-elasticsearch/schemabuilder/src/errors.rs

```diff
@@ -0,0 +1,37 @@
+use std::fmt;
+
+pub const UNSUPPORTED_MULTIPLE_OR_UNSPECIFIED_TYPES: &str =
+    "multiple non-trivial data types or unspecified data types are not supported";
+pub const UNSUPPORTED_NON_ARRAY_OR_OBJECTS: &str =
+    "data types other than objects or arrays of objects are not supported";
+pub const UNSUPPORTED_OBJECT_ADDITIONAL_FIELDS: &str =
+    "additional properties on an object are not supported";
+pub const UNSUPPORTED_TUPLE: &str = "Tuple is not supported";
+
+pub const POINTER_EMPTY: &str = "empty path";
+pub const POINTER_MISSING_FIELD: &str = "pointer of a non-existing field";
+pub const POINTER_WRONG_FIELD_TYPE: &str =
+    "non-leaf field on json path is a basic type (int, bool..)";
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("bad schema json")]
+    SchemaJsonParsing(#[from] serde_json::Error),
+    #[error("unsupported Flow schema in elastic search, details: {0}")]
+    UnSupportedError(String),
+    #[error("unable to override elastic search schema, details: {0}")]
+    OverridePointerError(String),
+}
+
+impl Error {
+    pub fn schema_error(error_message: &str, debug_comp: &dyn fmt::Debug) -> Self {
```

Comment:
Great. I've made the change accordingly. The only difference is  that `Shape` is actually referring to a few different structs, `Shape`, `ArrayShape`, etc. So I kept a Boxed `dyn fmt::Debug` to referring it. 

I am getting things working by using `box::new(shape.clone())`. The `clone` here seems to be unnecessary, however, if I remove it, the compiler will complain that it does not satisfy a `static lifetime requirement`. If there is any suggestions on improve, please let me know!

Reply:
There's a few different things that I can try to clear up here. 

First let's talk about functions like this one:

```rust
fn build_from_object<'a, 'b>(shape: &'a ObjShape) -> Result<ESFieldType, Error<'b>> {
    if !shape.additional.is_none() {
        return Err(Error::UnSupportedError {
            message: UNSUPPORTED_OBJECT_ADDITIONAL_FIELDS,
            shape: Box::new(shape.clone()),
        });
    }
    todo!("elided");
}
```

Here's the `shape` parameter is _borrowed_, which means that you have to "give it back" at the end of this function scope. But your `UnSupportedError` declares that it _owns_ its `shape` field. This is because its type is `Box<dyn fmt::Debug>`, which is an owned trait object. So the `.clone()` here is taking the variable you have (a `&'a ObjShape`) and converting it to an owned `ObjShape`, which is then converted to a boxed trait object when you call `Box::new`. Think of the `Box<dyn fmt::Eebug>` as equivalent to `Box<ObjShape>` with respect to ownership. In either case, `Box::new` is said to "take ownership" of whatever you pass to it.

It's actually a pretty common pattern to clone objects when constructing errors. This allows you to pass around borrows in the common happy case, while allowing your error types to also own their fields (it's generally best to keep lifetimes out of error types. More on that later).

So I think this function is actually fine, but hopefully you now understand why the `clone` was required.

Now let's talk about your `Error` type. Here I've elided the attributes:

```
pub enum Error<'a> {
    SchemaJsonParsing(serde_json::Error),
    UnSupportedError {
        message: &'static str,
        shape: Box<dyn fmt::Debug>,
    },
    OverridePointerError {
        message: &'static str,
        pointer: &'a str,
    },
}
```

The TLDR is that you should change `pointer` to an owned `String` and get rid of the generic lifetime parameter (`<'a>`). More generally, you should try to keep non-static borrows (i.e. `<'a>`) out of error types, cloning if necessary, because the lifetime parameter will prevent you from turning your error into a `Box<dyn std::error::Error>`, which is a really useful type to pass around. 

To elaborate, the compiler will not allow your `Error` values to live beyond the lifetime of the `pointer` that it borrows. **This is a good thing** but it does limit what you can do with the error. The `&'static` borrows are OK because the `'static` lifetime extends for the entire life of your program, so your Error type can't live any longer than that by definition. Most commonly, you'll use explicit `'static` lifetimes for string literals. You should read `Box<dyn Trait>` as syntactic sugar for `Box<dyn (Trait + 'static)>`, which helps make it more clear what's going on. You _could_ turn your `Error<'a>` into a `Box<Error<'a>>`, but the compiler will ensure that the box cannot outlive the thing that it borrows (the lifetime of `'a` becomes part of the Box). But the implicit `'static` lifetime on `Box<dyn Trait>` means that the compiler will consider it valid for instances of that type to live _forever_, which isn't OK, since the owner of `pointer` will presumably want to deallocate it at some point.

Going back to the functions like `fn build_from_object<'a, 'b>(shape: &'a ObjShape) -> Result<ESFieldType, Error<'b>>`, you should be able to remove the explicit lifetime annotations from this function once you remove the lifetime from `Error`. You probably put them in in the first place in order to tell the compiler that the lifetime parameter of the `Error` _isn't_ related to the lifetime of the `shape` borrow. This is because the compiler [infers](https://doc.rust-lang.org/reference/lifetime-elision.html) lifetime parameters for function signatures. Technically, you could have just returned `Error<'static>`, as long as you don't actually construct a variant with the `pointer`. But the best thing in this case is to remove the lifetime parameters.

Reply:
Yes, thank you @psFried for the detailed explanations. They are key to the point.

I was struggle about different ways of 
1. construct the `shape` earlier in the workflow to keep it the same lifetime as the `Error`, so that `shape` is valid to be borrowed from without boxing it. (just like `es_type_overrides`. 2. Using Boxed Shape. 3. Convert and store the error messages as Strings instead of `Shape`.
They all have sth. that I don't feel good about. And keeping all owned variables with`static` lifetime for errors sounds great.

I will look more and better understand the other points on the Box and Clone, and made changes as suggested.